src/openrct2/interface/window.c

```diff
@@ -856,6 +856,22 @@ void window_close_all_except_class(rct_windowclass cls)
     }
 }
 
+void window_close_all_except_flags(uint16 flags)
+{
+    for (rct_window * w = g_window_list; w < RCT2_NEW_WINDOW;)
+    {
+        if (!(w->flags & flags))
+        {
+            window_close(w);
+            w = g_window_list;
```

Comment:
`window_close` moves all the items following `w` and requires you to start over… if you looped over the windows in reverse order, you wouldn't have to restart and could reduce memory access.

Reply:
So like this?
```c
    for (rct_window * w = RCT2_NEW_WINDOW - 1; w >= g_window_list; w--)
    {
        if (!(w->flags & flags))
        {
            window_close(w);
        }
    }
```

Reply:
yes, although there is `RCT2_LAST_WINDOW` already.

Reply:
Oh, and while at it, please also change `windows_close_all` to this, I have just run into this bug that I mentioned above :P